### PR TITLE
Updated nfs.c to properly parse sharenfs args

### LINKED ISSUE #2773: Cannot specify different NFS share options for multiple hosts
I'm not sure if this is a bug or just my misunderstanding of how to do this, but it doesn't appear to be possible to specify different export options for multiple hosts for a single filesystem.  For example:

zfs set sharenfs='rw=@host1.home.net,no_subtree_check,no_root_squash,rw=@host2.home.net,no_subtree_check,root_squash,' data

I want my data filesystem to be exported to host1 w/ full read-write access, including root.  However, I want host2 to have read-write access only for standard users, NOT root.

The above sharenfs parameters do let me mount the data share from both host1 and host2, but I get root write access is denied on both host1 and host2.  It appears that the root_squash at the end of the list for host2 is overriding the no_root_squash option for host1, as if I change it to no_root-squash or just remove it altogether then both host1 and host2 can write.

I guess this makes sense sense the above parameter list is entirely comma-delimited (so how would ZFS know to apply the second set of options only to host2?), but I can't find any other way to make the share work.  I tried using a colon, semi-colon, parentheses, etc. to delimit the two hosts, but I get an "invalid options" message every time I try anything different.

Am I missing something obvious here, or is this a current limitation?  If the latter, is there any way to work around this (preferrably without switching to doing everything through /etc/exports, because I've already encountered limitations there that pushed me back to using ZFS' sharenfs :-) )?

Thanks.

### PR DESCRIPTION
**Pull request by dipack95**
Signed-off-by: Dipack P Panjabi <dpanjabi@hudson-trading.com>
Closes #2773

### Description

In the current iteration of the sharenfs option parsing,
it cannot distinguish between options set for different hosts.
Ex: sharenfs="rw=@host1,root_squash,rw=@host2,no_root_squash"
sets both hosts to NOT have their root user access squashed,
which is incorrect.

In this commit, I have just reproduced Turbo's (@FranzUrbo) work,
and rebased it against current master.

### How Has This Been Tested?

Tested by simply a running a command as follows:
```
zfs set sharenfs="rw=@host1,root_squash,rw=@host2,no_root_squash" pool1/d1
```
and checking the output of `exportfs -v`.

Tested on a Debian 9.9 VM, with the changes applied to commit 43a8536260e76dab4a615164f9e6d6397c6b7778 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by grwilson** [comment]
@dipack95 thank you for your PR submission. You may want to take a look at the [C style and coding standards](http://www.cis.upenn.edu/~lee/06cse480/data/cstyle.ms.pdf) as it seems like you have some extraneous changes which violate the standard. Can you take a look and update your PR?

**Comment by dipack95** [comment]
@grwilson My IDE seems to have applied it's default settings to format the code. Is there a recommended code formatter that I can use to bring the code style up to scratch?

**Comment by dipack95** [comment]
@grwilson I've fixed the formatting issues for the offending file.

**Comment by dipack95** [comment]
@grwilson @don-brady have you had a chance to review these changes?

File: lib/libshare/nfs.c [hunk 1]
```diff
@@ -323,19 +260,113 @@ add_linux_shareopt(char **plinux_opts, const char *key, const char *value)
...
 
-	/* host-specific options, these are taken care of elsewhere */
-	if (strcmp(key, "ro") == 0 || strcmp(key, "rw") == 0 ||
-	    strcmp(key, "sec") == 0)
-		return (SA_OK);
+// TODO: Make sure that a 'sec=yyy' (or any other option) BEFORE a/any
+//  'rw[=@xxx]' or 'ro[=@xxx]' works.
+	if (strcmp(key, "ro") == 0 || strcmp(key, "rw") == 0) {
+
+		opts = (nfs_share_list_t *)malloc(sizeof (nfs_share_list_t));
```
Comment (freqlabs) [inline lib/libshare/nfs.c]: I don't see where all these allocations are being freed. Is it expected for the library to leak memory?

File: lib/libshare/nfs.c [hunk 2]
```diff
@@ -47,10 +54,41 @@ static sa_fstype_t *nfs_fstype;
...
 typedef int (*nfs_host_callback_t)(const char *sharepath, const char *host,
-    const char *security, const char *access, void *cookie);
+	const char *security, const char *access,
+	void *cookie);
+
+/* List of hosts and their options */
+/* NOTE: share path is stored elsewhere */
+typedef struct nfs_share_list_s {
+	char host[255];
+	char opts[255];
```
Comment (freqlabs) [inline lib/libshare/nfs.c]: I don't see where the length of input strings are being validated. How do we protect against writing past the end of these arrays and corrupting nearby memory?

**Comment by dascionepppl** [comment]
Any update on this? We have a fairly large deployment that we need this functionality for and it seems like this issue is pretty old

**Comment by ahrens** [comment]
I agree that this is a problem, but this might not be the right way to address it.  We might instead remove the parsing code from ZFS, and just pass the string down to the OS-specific parser (as @grwilson suggested elsewhere).  Alternatively, we could simplify the parser to a bare minimum that we can parse simply and correctly, and translate that to the OS-specific flags.

**Pull request closed without merging**